pkg/signature/kms/kms.go

```diff
@@ -50,11 +55,25 @@ var providersMap = map[string]ProviderInit{}
 // Get returns a KMS SignerVerifier for the given resource string and hash function.
 // If no matching provider is found, Get returns a ProviderNotFoundError. It
 // also returns an error if initializing the SignerVerifier fails.
+// If keyResourceID doesn't match any of our hard-coded provers' schemas,
+// it will try to use the plugin system as a provider. If the caller did not import cliplugin
+// to allow it to run its init(), then it returns ProviderNotFoundError.
 func Get(ctx context.Context, keyResourceID string, hashFunc crypto.Hash, opts ...signature.RPCOption) (SignerVerifier, error) {
 	for ref, pi := range providersMap {
 		if strings.HasPrefix(keyResourceID, ref) {
-			return pi(ctx, keyResourceID, hashFunc, opts...)
+			sv, err := pi(ctx, keyResourceID, hashFunc, opts...)
+			if err != nil {
+				return nil, err
+			}
+			return sv, nil
+		}
+	}
+	if pi, ok := providersMap[CLIPluginProviderKey]; ok {
+		sv, err := pi(ctx, keyResourceID, hashFunc, opts...)
```

Comment:
Is there a reason to limit the system to one registered plugin at a time? This seems like a change compared to current setup (assuming we at some point follow the plan of implementing current KMS providers as plugins).

At first glance I don't see why CLIPluginProviderKey exists -- could the plugins use the plugin name as the argument to AddProvider, and then this code could do the same thing the existing provider matching does: if the keyResourceID has a prefix that matches the provider key, use that provider?

Reply:
You can actually have many plugin programs on the system, and you can choose which plugin program based on the `keyResourceID`: e.g., `my-hsm:///my-key` or `digitalocean]://my-key`. 

What's "registered" is actually only the provider's initialization function that produces a `SignerVerifier` implementation. `CLIPluginProviderKey` only exists because we need a way to access `clipugin`'s the initialization function, but we cannot import it from the `kms` package because that would cause a circular import, since `cliplugin` already imports `kms`. So, when `cliplugin` is imprted, its package `init()` will call `AddProvider(CLIPluginProviderKey, signer_verifier_maker_func)`.

> could the plugins use the plugin name as the argument to AddProvider

Yes, but that would mean that the main program has to first discover all potential plugin programs, and then register each of their initialization functions. I thought it simpler to just first determine of the caller intended to use a plugin or not. Then, determining which plugin to use is only a matter of parsing the keyResourceID and finding the corresponding program on the system.

Eventually when all the providers become plugins, we won't need this registration process. But for now, we need to keep the diff small and not break compatibility.

Reply:
> You can actually have many plugin programs on the system

do you mean you can have multiple binaries installed on the system? That seems correct, but a single application can only load one plugin, correct? Or more precisely, only the last loaded plugin is actually used. 

Now this may be as designed but 
* I didn't understand this from the design doc so wanted to make sure
* this seems like a change compared to current KMS implementation: I think you can currently use multiple providers (unsure if anyone is interested in this)

Reply:
I can test this later, but: A single application should be be able to load many plugins simultaneously via `Get()`. The returned `PluginClient` doesn't modify any global variables, so each instance should be independent.

As for many instances of the same PluginClient that all talk to the same plugin program, it would be up to the authors to make their plugins be able to handle multiple simultaneous CLI invocations (like TUF caching).

Reply:
Get() looks up the plugin provider with a constant key. I don't see how it could return different plugin providers

Reply:
`providersMap[CLIPluginProviderKey]` returns a function to load a SignerVerifier, and it isn't caching that result anywhere. If you invoke `Get()` repeatedly with different keys, I believe it should create multiple `SignerVerifier` clients.